File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        unordered_map<int, vector<int>> adj;
        
        for(auto vec: prerequisites){
            int v = vec[0];
            int u = vec[1];
            
            indegree[v]++;

            adj[u].push_back(v);
        }

        queue<int> q;
        int count = 0;

        for(int i=0; i<numCourses; i++){
            if(indegree[i] == 0){
                q.push(i);
                count++;
            }
        }
        
        while(!q.empty()){
            int u = q.front();
            q.pop();
            for(auto &v: adj[u]){
                indegree[v]--;
                if(indegree[v] == 0){
                    q.push(v);
                    count++;
                }
            }
        }
        
        if(count == numCourses){
            return true;
        }

        return false;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj_graph = [[] for i in range(numCourses)]
        in_degree = [0] * numCourses
        for i in range(len(prerequisites)):
            dest, source = prerequisites[i]
            adj_graph[source].append(dest)
            in_degree[dest] += 1
        
        courses = 0
        q = deque()
        for i in range(numCourses):
            if in_degree[i] == 0:
                q.append(i)
                courses += 1
        
        while q:
            current_source = q.pop()
            for edge in adj_graph[current_source]:
                in_degree[edge] -= 1
                if in_degree[edge] == 0:
                    q.append(edge)
                    courses += 1
        
        return courses == numCourses
        